File: LLD/Design Parking Lot/parking_lot.py
import logging
from typing import List, Dict, Optional
from dataclasses import dataclass
from collections import defaultdict
from abc import ABC, abstractmethod

# Configure logging
logging.basicConfig(level=logging.INFO, format='%(asctime)s - %(levelname)s - %(message)s')

@dataclass
class Vehicle:
    vehicle_type: int      # 2 or 4 wheeler
    vehicle_number: str    # Vehicle registration
    ticket_id: str         # Unique ticket
    spot_id: str           # Where it's parked

# Strategy Pattern: Abstract base class for parking strategies
class ParkingStrategy(ABC):
    @abstractmethod
    def find_spot(self, parking_lot, vehicle_type: int) -> Optional[str]:
        """Find a parking spot based on the strategy"""
        pass

# Concrete Strategy: Find first available spot
class FirstAvailableSpotStrategy(ParkingStrategy):
    def find_spot(self, parking_lot, vehicle_type: int) -> Optional[str]:
        """Find first available spot (lowest indices)"""
        for floor in range(parking_lot.floors):
            for row in range(parking_lot.rows):
                for col in range(parking_lot.cols):
                    if (parking_lot.parking[floor][row][col] == vehicle_type and 
                        parking_lot._get_spot_id(floor, row, col) not in parking_lot.spot_map):
                        return parking_lot._get_spot_id(floor, row, col)
        return None

# Concrete Strategy: Find spot on floor with maximum free spots
class MaxFreeSpotsFloorStrategy(ParkingStrategy):
    def find_spot(self, parking_lot, vehicle_type: int) -> Optional[str]:
        """Find spot on floor with maximum free spots"""
        max_spots = -1
        chosen_floor = -1
        
        # Find floor with maximum free spots
        for floor in range(parking_lot.floors):
            free_count = parking_lot.free_spots[floor][vehicle_type]
            if free_count > max_spots:
                max_spots = free_count
                chosen_floor = floor
        
        if chosen_floor == -1:
            return None
            
        # Find first available spot on chosen floor
        for row in range(parking_lot.rows):
            for col in range(parking_lot.cols):
                if (parking_lot.parking[chosen_floor][row][col] == vehicle_type and 
                    parking_lot._get_spot_id(chosen_floor, row, col) not in parking_lot.spot_map):
                    return parking_lot._get_spot_id(chosen_floor, row, col)
        return None

class ParkingLot:
    def __init__(self, parking: List[List[List[int]]]):
        self.parking = parking
        self.floors = len(parking)
        self.rows = len(parking[0])
        self.cols = len(parking[0][0])
        
        # Maps to track vehicle locations and spots
        self.vehicle_map: Dict[str, Vehicle] = {}
        self.spot_map: Dict[str, Vehicle] = {}
        
        # Initialize counts of free spots per floor and type
        self.free_spots = defaultdict(lambda: defaultdict(int))
        self._initialize_free_spots()
        
        # Initialize parking strategies
        self.strategies = {
            0: FirstAvailableSpotStrategy(),
            1: MaxFreeSpotsFloorStrategy()
        }
    
    def _initialize_free_spots(self):
        """Initialize the count of free spots for each floor and vehicle type"""
        for floor in range(self.floors):
            for row in range(self.rows):
                for col in range(self.cols):
                    spot_type = self.parking[floor][row][col]
                    if spot_type in (2, 4):  # Active spots only
                        self.free_spots[floor][spot_type] += 1
    
    def _get_spot_id(self, floor: int, row: int, col: int) -> str:
        """Generate spot ID from coordinates"""
        return f"{floor}-{row}-{col}"
    
    def park(self, vehicle_type: int, vehicle_number: str, ticket_id: str, 
            parking_strategy: int) -> str:
        """Park a vehicle using the specified strategy"""
        # Check if vehicle is already parked
        if vehicle_number in self.vehicle_map or ticket_id in self.vehicle_map:
            logging.warning(f"Vehicle {vehicle_number} or ticket {ticket_id} is already parked.")
            return ""
            
        # Get the appropriate strategy and find spot
        strategy = self.strategies.get(parking_strategy)
        if not strategy:
            logging.error("Invalid parking strategy.")
            return ""
            
        spot_id = strategy.find_spot(self, vehicle_type)
        if not spot_id:
            logging.info(f"No available spot for vehicle type {vehicle_type}.")
            return ""
            
        # Create and store vehicle information
        vehicle = Vehicle(vehicle_type, vehicle_number, ticket_id, spot_id)
        self.vehicle_map[vehicle_number] = vehicle
        self.vehicle_map[ticket_id] = vehicle
        self.spot_map[spot_id] = vehicle
        
        # Update free spots count
        floor = int(spot_id.split("-")[0])
        self.free_spots[floor][vehicle_type] -= 1
        
        logging.info(f"Vehicle {vehicle_number} parked at {spot_id}.")
        return spot_id
    
    def remove_vehicle(self, spot_id: str) -> bool:
        """Remove a vehicle from its parking spot"""
        if spot_id not in self.spot_map:
            logging.warning(f"Spot ID {spot_id} not found.")
            return False
            
        vehicle = self.spot_map[spot_id]
        floor = int(spot_id.split("-")[0])
        
        # Update maps and free spots count
        del self.vehicle_map[vehicle.vehicle_number]
        del self.vehicle_map[vehicle.ticket_id]
        del self.spot_map[spot_id]
        self.free_spots[floor][vehicle.vehicle_type] += 1
        
        logging.info(f"Vehicle {vehicle.vehicle_number} removed from {spot_id}.")
        return True
    
    def search_vehicle(self, query: str) -> str:
        """Search for a vehicle by number or ticket ID"""
        if query in self.vehicle_map:
            spot_id = self.vehicle_map[query].spot_id
            logging.info(f"Vehicle {query} found at {spot_id}.")
            return spot_id
        logging.warning(f"Vehicle {query} not found.")
        return ""
    
    def get_free_spots_count(self, floor: int, vehicle_type: int) -> int:
        """Get count of free spots for a vehicle type on a floor"""
        return self.free_spots[floor][vehicle_type]

# Example usage
def main():
    parking = [[[4, 4, 2, 2],
                [2, 4, 2, 0],
                [0, 2, 2, 2],
                [4, 4, 4, 0]]]
    
    lot = ParkingLot(parking)
    
    # Test parking
    spot1 = lot.park(2, "bh234", "tkt4534", 0)
    spot2 = lot.park(4, "bh235", "tkt4535", 1)
    
    # Test searching
    lot.search_vehicle("bh234")
    
    # Test removal
    lot.remove_vehicle(spot1)

if __name__ == "__main__":
    main()


"""
@dataclass
class Vehicle:
    vehicle_type: int      # 2 or 4 wheeler
    vehicle_number: str    # Vehicle registration
    ticket_id: str         # Unique ticket
    spot_id: str           # Where it's parked

class ParkingStrategy(ABC):
    @abstractmethod
    def find_spot(self, parking_lot, vehicle_type: int) -> Optional[str]:
        pass

class FirstAvailableSpotStrategy(ParkingStrategy):
    def find_spot(self, parking_lot, vehicle_type: int) -> Optional[str]:
        pass
        
class MaxFreeSpotsFloorStrategy(ParkingStrategy):
    def find_spot(self, parking_lot, vehicle_type: int) -> Optional[str]:
        pass

class ParkingLot:
    def __init__(self, parking):
        pass
    def park(self, vehicle_type: int, vehicle_number: str, ticket_id: str, 
            parking_strategy: int) -> str:
        pass
    def remove_vehicle(self, spot_id: str) -> bool:
        pass
    def search_vehicle(self, query: str) -> str:
        pass
    def get_free_spots_count(self, floor: int, vehicle_type: int) -> int:
        pass
    
"""

File: LLD/Design Parking Lot/ParkingLot.cs
using System;
using System.Collections.Generic;

namespace ParkingLotSystem
{
    // Vehicle class to represent a parked vehicle
    public class Vehicle
    {
        public int VehicleType { get; set; } // 2 or 4 wheeler
        public string VehicleNumber { get; set; } // Vehicle registration
        public string TicketId { get; set; } // Unique ticket
        public string SpotId { get; set; } // Where it's parked

        public Vehicle(int vehicleType, string vehicleNumber, string ticketId, string spotId)
        {
            VehicleType = vehicleType;
            VehicleNumber = vehicleNumber;
            TicketId = ticketId;
            SpotId = spotId;
        }
    }

    // Strategy Pattern: Abstract base class for parking strategies
    public interface IParkingStrategy
    {
        string FindSpot(ParkingLot parkingLot, int vehicleType);
    }

    // Concrete Strategy: Find first available spot
    public class FirstAvailableSpotStrategy : IParkingStrategy
    {
        public string FindSpot(ParkingLot parkingLot, int vehicleType)
        {
            for (int floor = 0; floor < parkingLot.Floors; floor++)
            {
                for (int row = 0; row < parkingLot.Rows; row++)
                {
                    for (int col = 0; col < parkingLot.Cols; col++)
                    {
                        if (parkingLot.Parking[floor][row][col] == vehicleType &&
                            !parkingLot.SpotMap.ContainsKey(parkingLot.GetSpotId(floor, row, col)))
                        {
                            return parkingLot.GetSpotId(floor, row, col);
                        }
                    }
                }
            }
            return null;
        }
    }

    // Concrete Strategy: Find spot on floor with maximum free spots
    public class MaxFreeSpotsFloorStrategy : IParkingStrategy
    {
        public string FindSpot(ParkingLot parkingLot, int vehicleType)
        {
            int maxSpots = -1;
            int chosenFloor = -1;

            for (int floor = 0; floor < parkingLot.Floors; floor++)
            {
                int freeCount = parkingLot.FreeSpots[floor][vehicleType];
                if (freeCount > maxSpots)
                {
                    maxSpots = freeCount;
                    chosenFloor = floor;
                }
            }

            if (chosenFloor == -1) return null;

            for (int row = 0; row < parkingLot.Rows; row++)
            {
                for (int col = 0; col < parkingLot.Cols; col++)
                {
                    if (parkingLot.Parking[chosenFloor][row][col] == vehicleType &&
                        !parkingLot.SpotMap.ContainsKey(parkingLot.GetSpotId(chosenFloor, row, col)))
                    {
                        return parkingLot.GetSpotId(chosenFloor, row, col);
                    }
                }
            }
            return null;
        }
    }

    // ParkingLot class to manage the parking system
    public class ParkingLot
    {
        public int Floors { get; }
        public int Rows { get; }
        public int Cols { get; }
        public int[][][] Parking { get; }
        public Dictionary<string, Vehicle> VehicleMap { get; }
        public Dictionary<string, Vehicle> SpotMap { get; }
        public Dictionary<int, Dictionary<int, int>> FreeSpots { get; }
        public Dictionary<int, IParkingStrategy> Strategies { get; }

        public ParkingLot(int[][][] parking)
        {
            Parking = parking;
            Floors = parking.Length;
            Rows = parking[0].Length;
            Cols = parking[0][0].Length;

            VehicleMap = new Dictionary<string, Vehicle>();
            SpotMap = new Dictionary<string, Vehicle>();
            FreeSpots = new Dictionary<int, Dictionary<int, int>>();

            Strategies = new Dictionary<int, IParkingStrategy>
            {
                { 0, new FirstAvailableSpotStrategy() },
                { 1, new MaxFreeSpotsFloorStrategy() }
            };

            InitializeFreeSpots();
        }

        private void InitializeFreeSpots()
        {
            for (int floor = 0; floor < Floors; floor++)
            {
                FreeSpots[floor] = new Dictionary<int, int>
                {
                    { 2, 0 },
                    { 4, 0 }
                };

                for (int row = 0; row < Rows; row++)
                {
                    for (int col = 0; col < Cols; col++)
                    {
                        int spotType = Parking[floor][row][col];
                        if (spotType == 2 || spotType == 4)
                        {
                            FreeSpots[floor][spotType]++;
                        }
                    }
                }
            }
        }

        public string GetSpotId(int floor, int row, int col)
        {
            return $"{floor}-{row}-{col}";
        }

        public string Park(int vehicleType, string vehicleNumber, string ticketId, int parkingStrategy)
        {
            if (VehicleMap.ContainsKey(vehicleNumber) || VehicleMap.ContainsKey(ticketId))
            {
                Console.WriteLine($"Vehicle {vehicleNumber} or ticket {ticketId} is already parked.");
                return null;
            }

            var strategy = Strategies[parkingStrategy];
            string spotId = strategy.FindSpot(this, vehicleType);
            if (spotId == null)
            {
                Console.WriteLine($"No available spot for vehicle type {vehicleType}.");
                return null;
            }

            var vehicle = new Vehicle(vehicleType, vehicleNumber, ticketId, spotId);
            VehicleMap[vehicleNumber] = vehicle;
            VehicleMap[ticketId] = vehicle;
            SpotMap[spotId] = vehicle;

            int floor = int.Parse(spotId.Split('-')[0]);
            FreeSpots[floor][vehicleType]--;

            Console.WriteLine($"Vehicle {vehicleNumber} parked at {spotId}.");
            return spotId;
        }

        public bool RemoveVehicle(string spotId)
        {
            if (!SpotMap.ContainsKey(spotId))
            {
                Console.WriteLine($"Spot ID {spotId} not found.");
                return false;
            }

            var vehicle = SpotMap[spotId];
            int floor = int.Parse(spotId.Split('-')[0]);

            VehicleMap.Remove(vehicle.VehicleNumber);
            VehicleMap.Remove(vehicle.TicketId);
            SpotMap.Remove(spotId);
            FreeSpots[floor][vehicle.VehicleType]++;

            Console.WriteLine($"Vehicle {vehicle.VehicleNumber} removed from {spotId}.");
            return true;
        }

        public string SearchVehicle(string query)
        {
            if (VehicleMap.ContainsKey(query))
            {
                string spotId = VehicleMap[query].SpotId;
                Console.WriteLine($"Vehicle {query} found at {spotId}.");
                return spotId;
            }
            Console.WriteLine($"Vehicle {query} not found.");
            return null;
        }

        public int GetFreeSpotsCount(int floor, int vehicleType)
        {
            return FreeSpots[floor][vehicleType];
        }
    }

    // Example usage
    class Program
    {
        static void Main(string[] args)
        {
            int[][][] parking = new int[][][]
            {
                new int[][] 
                {
                    new int[] { 4, 4, 2, 2 },
                    new int[] { 2, 4, 2, 0 },
                    new int[] { 0, 2, 2, 2 },
                    new int[] { 4, 4, 4, 0 }
                }
            };

            var lot = new ParkingLot(parking);

            // Test parking
            string spot1 = lot.Park(2, "bh234", "tkt4534", 0);
            string spot2 = lot.Park(4, "bh235", "tkt4535", 1);

            // Test searching
            lot.SearchVehicle("bh234");

            // Test removal
            lot.RemoveVehicle(spot1);
        }
    }
}
